lexer.py

```diff
@@ -0,0 +1,127 @@
+        self.tokens.append(Token(token_type, self.current_char, self.line))
+        self._advance()
+    def _add_token2(self, token_type):
+        self.tokens.append(Token(token_type, self.current_char + self._peek(), self.line))
+        self._advance()
+        self._advance()
+    def _id(self):
+        result = ""
+        while self.current_char is not None and (self.current_char.isalnum() or self.current_char == "_"):
+            result += self.current_char
+            self._advance()
+        if result == "TRUE": token = Token(TokenTypes.BOOL, True, self.line)
+        elif result == "FALSE": token = Token(TokenTypes.BOOL, False, self.line)
+        elif result in ("AND", "OR", "NOT"): token = Token(TokenTypes.OP, result, self.line)
+        else: token = Token(TokenTypes.KW if result in RESERVED_WORDS else TokenTypes.ID, result, self.line)
+        self.tokens.append(token)
+        self._eat_white_space()
+    def _str(self):
+        result = ""
+        while self.current_char is not None and self.current_char != "'":
+            if self.current_char == "\\":
+                escape_char = ESCCHARS.get(self._peek(), None)
+                if escape_char is None:
+                    self._error("invalid escape sequence")
+                result += escape_char
+                self._advance()
+                self._advance()
+            else:
+                result += self.current_char
+                self._advance()
+        self._advance()
+        token = Token(TokenTypes.STR, result, self.line)
+        self.tokens.append(token)
+        self._eat_white_space()
+    def _num(self):
+        result = ""
+        while self.current_char is not None and self.current_char.isdigit():
+            result += self.current_char
+            self._advance()
+        if self.current_char == ".":
+            result += self.current_char
+            self._advance()
+            while self.current_char is not None and self.current_char.isdigit():
+                result += self.current_char
+                self._advance()
+        token = Token(TokenTypes.NUM, float(result), self.line)
+        self.tokens.append(token)
+        self._eat_white_space()
+    def lex(self):
+        # tokenizer
+        text = self.text
+        while self.pos < len(text):
+            while self.current_char in (" ", "\t", "\n"):
+                self._advance()
+            if self.pos >= len(text):
+                self.tokens.append(Token(TokenTypes.EOF, None, self.line))
+            elif self.current_char == "-" and self._peek() == "-":   # inline comments
+                while self.current_char != "\n" and self.current_char != None:
+                    self._advance()
+            elif self.current_char == "'":                          # strings
+                self._advance()
+                self._str()
+            elif self.current_char.isdigit() or self.current_char == '.':                       # numbers
```

Comment:
Nitpick: why are these comments (`# inline comments`, `# strings`, `# numbers`) aligned so strangely?  A single space after the colon should be enough.